README.md
[comment]: <> ([![Travis-CI Build Status](https://travis-ci.org/.svg?branch=master)](https://travis-ci.org/)[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/johncheetah/fbadsinsightsr?branch=master&svg=true)](https://ci.appveyor.com/project/johncheetah/fbadsinsightsr)[![Coverage Status](https://img.shields.io/codecov/c/github//master.svg)](https://codecov.io/github/?branch=master)[![Coverage Status](https://img.shields.io/coveralls/.svg)](https://coveralls.io/r/?branch=master))
![Developped by Cheetah Mobile](http://www.mobyaffiliates.com/wp-content/uploads/2015/10/12.png)

# fbAdsInsightsR #

Current version: **v3.1** (*see "Developments" section below*)

fbAdsInsightsR is an R package that allows fetching data from the [facebook Ads Insights API](https://developers.facebook.com/docs/marketing-api/insights/v2.7). 
**Now updated to Facebook Graph API v2.7!**

## Functions ##


- OAuth
    * `fbAuthenticate`
- GET-family
    * `getAccount`
    * `getCampaign`
    * `getAdset`
    * `getAd`
    * `getAny`
    * `getActivity`
    * `getLabCampaigns` 
    * `getLabAdsets`
    * `getLabAds`
    * `getLabCreatives`
    * `getEstreach` (dev)
- List-family
    * `listBusinesses`
    * `listAccounts`
    * `listVideos`
    * `listImages`
    * `listCreatives`
    * `listApps`
    * `listAdApps` 
    * `ListLabels` 
- Find-family
    * `findFields`
    * `findParams`
- Grab-family
    * `grabCampaigns`
    * `grabAdsets`
    * `grabAds`
    * `grabCreatives` (.deprecated for `listCreatives`)
- Check-family
    * `checkTarget` (.deprecated for `checkTargetSentence`)
    * `checkStatus`
    * `checkTargetSentence` 
    * `checkTargetTree`
    * `checkUsers` (dev)
- Update-family
    * `updateCampaign`
    * `updateAdset` 
    * `updateAd`
- Create-family
    * `createLabels` 
    
## Issues & Feature Requests

Please report bugs and request features on the [issue page](https://bitbucket.org/JohnCheetah/fbadsinsightsr/issues).

## Documentation ##

Constantly being updated and improved. Currently the manual and examples are available in own repository (access NOT restricted) [here](https://bitbucket.org/JohnCheetah/fbadsinsightsrdocs/src) Alternatively refer to the list below.

* [Manual](https://bitbucket.org/JohnCheetah/fbadsinsightsr/downloads)
* [Vignette](https://bitbucket.org/JohnCheetah/fbadsinsightsrdocs/src) (v3.0 - update to *v3.1* WIP)
* [Inner-workings](https://bitbucket.org/JohnCheetah/fbadsinsightsr/downloads) - technical details (for devs).

## Install ##

Since access to the repository is restricted you will need your login and password to install the package.

`devtools::install_bitbucket("JohnCheetah/fbadsinsightsr", auth_user = "your_login", password = "your_password", ref = "Rcpp")`

## Examples ##

```R
# run authentication with your app details
TK <- fbAuthenticate(app.id = "1234567890123456", app.secret = "16xx79321xx0130x2x10a08x3e2x80xx", scope = c("ads_management", "ads_read"))
                           
accounts <- listAccounts(id = "me", token = TK) # list facebook advertising accounts you have access to

ads <- grabAds(sample(accounts$id, 1), TK) # grab list of ads under random account

# get data on random ad
set.seed(19880525)
ad_data <- getAd(ad.id = sample(ads$id, 1), token = TK)

# get daily performance data for last month with a summary
my_preset <- findParams("date.preset")[grep("last_month", findParams("date.preset"))]
ad_data <- getAd(ad.id = sample(ads$id, 1), token = TK, n = 1000, date.preset = my_preset, time.increment = 1)
                
# get ad performance data by country 
ad_data <- getAd(ad.id = sample(ads$id, 1), token = TK, breakdowns = "country")
                  
# get ad performance data by age and gender 
ad_data <- getAd(ad.id = sample(ads$id, 1), token = TK, breakdowns = c("age", "gender"))
                
creatives <- listCreatives(sample(accounts$id, 1), TK) # list creatives

# check targeting of adset
adsets <- grabAdsets(sample(accounts$id, 1), TK)
(target <- checkTargetSentence(sample(adsets$id, 1), TK))

targeting <- checkTargetTree(sample(accounts$id, 1), TK) # check targeting from account

labels <- listLabels(accounts$id[1], TK) # get labels

# get ads that match ANY label
ad_labels <- getLabAds(accounts$id[1], labels = labs, operator = "ANY", token = TK)

# get campaigns that match ALL labels                        
ad_labels <- getLabAds(accounts$id[1], labels = labs, operator = "ALL", token = TK)
                      
(apps <- listApps(accounts$id[1], TK)) # list apps
(ad_apps <- listAdApps(accounts$id[1], TK)) # list advertiseable apps
```

## Developments ##

### Versions ###

See Git tags: *2.0*, *v3.0* and *v3.1* (Rcpp branch)

### Dev ###

- New functions: `checkUsers`, `getEstReach` 
- `listAccounts` id parameter defaults to `me`
- fixed return of `getAny` FUN where it sometimes would not return results when few fields were requested, also fixed sometimes returning *weirdly* named variables.
- Get-family functions now compute various ratios CVR, CPI, CPL, and CPA.
- `breakdown` parameter `placement_rhc` has been removed in API v2.7

See NEWS.md for entire changelog.

### Branches ###

- **master** - Last release or dev version
- **Rcpp** - Uses C++ to speed up certain functions as well as v2.7 of API

## R CMD check

R CMD check results
`0 errors | 0 warnings | 0 notes`

## Access & Contributors ##

Access to the package is restricted, email John Coene & GaoCong if you want to grant access to more users.

* John Coene <john.coene@cmcm.com> (Author & maintainer)
* DongYuNan <dongyunan@cmcm.com> (User & tester)
* GaoCong <gaocong1@cmcm.com> (User)

---------------------------------------------------------------------------

by Cheetah Mobile - 2016

NEWS.md
## Rcpp dev ##

#### Updated to *v2.7*

- New function `checkUsers`, returns users and access permissions 
- `listAccounts` id parameter defaults to `me`
- fixed return of `getAny` FUN where it sometimes would not return results when few fields were requested, also fixed sometimes returning *weirdly* named variables.
- New function `getEstReach`, returns estimated reach
- Get-family functions now compute various ratios CVR, CPI, CPL, and CPA.
- `breakdown` parameter `placement_rhc` has been removed in API v2.7
- `getActivity` variable names returned corrected 

## v3.0 ##

Version Three brings many changes. The package now calls the updated **v2.6** of the API.

Changes:

1. `checkTarget` has been deprecated in favour of:
    1. `checkTargetSentence` AND 
    2. `checkTargetTree` which have been added, both retrieve targeting specifications but return the data in different formats and can be called on different objects (accounts, campaigns, adsets, ads). See documentation for more details.
2. `grabAccounts` and `grabImages` have been deprecated in favour of `listAccounts` and `listImages`
3. New function `listApps` returns apps under account
4. New function `listAdApps` returns list of advertiseable applications
5. `findFields` now also applies to `getLabCampaigns`, `getLabAdsets` and `getLabAds`
6. `findParams` now also applies to `limit.type` from new FUN `checkTargetTree`
7. New label-related functions:
    1. `listLabels` returns labels under account
    2. `getLabCampaigns` fetch campaigns by labels
    3. `getLabAdsets` fetch adsets by labels
    4. `getLabAds` fetch ads by labels
    5. `getLabCreatives` fetch creatives by labels
    6. `createLabels` create labels
8. New update-family functions:
    1. `updateCampaign`
    2. `updateAdset`
    3. `updateAd`
9. Number of fields for get-family extended from 74 to 90.

Bug fixes:

1. OAuth function now properly passes scope, multiple scopes can be passed.
2. `limit` parameter and check-family functions now works correctly. 

## v2.0 ##

Package overhauled; now internally functions on new S3 class for robustness, much optimisation and clean up of `utils.R`, package now split into 4 families of functions (`grab`, `check`, `find` and `get`), and much more.

See updated [documentation](https://bitbucket.org/JohnCheetah/fbadsinsightsrdocs/src) for details. 

* `fbAutenticate` FUN fixed

### v1.1 ###

* Data parsing fixed at last. Now dataframe returned by GET-family functions properly displays `actions`, `unique_actions`, `cost_per_action_type`, `cost_per_unique_action_type` and `website_ctr` as respective variables (column names) regardless of parameters or fields passed to functions.
* The latter improvement means the `simplify` argument is now irrelevant and has thus been deprecated.
* Sleep when `n > 100` in GET-family functions and `findInfo` has been reduced from 3 seconds to 0.5 seconds.
* Overall clean up in `internal.R` speeds up some processes.
* `n` and `verbose` added to `findAccounts`.
* Modified `verbose = TRUE` output to something more sensible.
* `findStatus` function added, returns status of specific campaign, adset or ad.
* `findCampaigns`, `findAdsets` and `findAds` functions added.
* `findInfo` now call `findObjects`
* `getCreative` added to fetch ad Creatives from ads.
* Arguments of functions appear in correct order in order to ease simple queries.
* Check for presence of data and `warning` returned now more adequate
* Overall clean up of GET-family and find-family functions.

Bug Fixes

* `paginate` (internal funciton) no longer infinitely loops.
* `toDF` (internal function) properly parse missing data.
* typos corrected (mainly in examples)
* Number of results correctly returned when `verbose = TRUE` 
* fixed `toDF` (internal.R) - GET-family should no longer return errors.
* `verbose = TRUE` has been fixed where it was not being paste o nnew line when used inside a loop

### v1.0 ###

* Temporary fix to mess in return from GET-family functions; using `simplify = TRUE` in `getAny` will ignore the fields that cause the issue.
* Added `getImage` function to fetch URLs of images used by an ad account, adset, etc.
* `findAccounts` also returns account name.
* Updated manual

Bug Fixes:

* In GET-family functions added `stop` message when using region as `breakdowns` together with `action_carousel_card_id` and/or `action_carousel_card_name` as fields as it is not allowed by API and returned an error. The latter two fields have also been removed when using `simplify=TRUE`.
* Fixed error when querying little data using GET-family function, was caused by `stop` message when no data - has been replaced by `warning`.

### v0.9 ###

* Simplifies data returned by GET-famlily functions if possible, though rarely the case. This is first attempt at dealing with the actions-related data which produce confusing dataframes.
* `paginate` function (`utils.R`) to clean up GET-family functions.
* `verbose` arguement now also returns details on pagination.
* Some clean up in both find and GET family functions.
* Manual Updated.

Bug Fixes:

* Fixed bug in the recently added `findTarget` function where the presence of an argument broke the function - should no longer occur.
* Fixed `findTarget` where `NULL` was returned when few specifications were retrieved.
* `findTarget` now correctly returns targeting specs.
* `findAccount` is no longer "missing verbose"

### v0.8 ###

* Added `verbose` arguement to GET-family functions. See documentation.
* Added `findTarget` function to retrieve targeting specs of an ad or adset.
* `findAccounts` function added; retrieves all account IDs accessible by either a business.facebook.com or a user. See updated Manual or ?documentation. 

Bug fixes:

* Fixed minor bug amongst GET-family functions where some errors on the inital queries where not caught if the response had little content.


R/getAny.R
#' Get data on any object
#' 
#' @description Get insights on advertising performances of an ad, an adset, a campaign or even an account. This function may relace any other GET function in this package.
#' 
#' @param id The id of the object you want to retrieve (Required),
#'  see \href{https://www.facebook.com/business/help/1492627900875762}{how to find yours}.
#' @param fields 
#' There are in total 73 valid fields defaults to 
#' (\code{default}) which returns the most popular ones. 
#' Run \code{\link{findFields}} to see all valid fields.
#' @param action.attribution.windows 
#' Determines what is the attribution window for the actions.
#'  For example, \code{c("28d_click")} means the API returns all actions
#'   that happened 28 days after someone clicked on the ad (Optional).
#'    See details below for valid values.
#' @param action.breakdowns 
#' How to break down action results. Supports more than one breakdowns 
#' (Optional). Run \code{\link{findParams}} to see all 
#' valid action breakdowns.
#' @param action.report.time 
#' Determines the report time of action stats. 
#' For example, if a person saw the ad on Jan 1st but converted on Jan 2nd,
#'  when you query the API with \code{action.report.time="impression"}, 
#'  you will see a conversion on Jan 1st. When you query the API with 
#'  \code{action.report.time="conversion"}, you will see a conversion 
#'  on Jan 2nd (Optional).
#' @param breakdowns
#'  How to break down the result. Does not support more than one breakdown,
#'   except \code{c("age", "gender")} and 
#'   \code{"impression_device", "placement"}. The option 
#'   \code{impression_device} cannot be used by itself (Optional). Run 
#'   \code{\link{findParams}} to see all valid breakdowns.
#' @param date.preset
#'  Represents a relative time range (Optional). This field is ignored if 
#'  \code{time.range} is specified. Run \code{\link{findParams}} 
#'  to see all valid presets.
#' @param level
#'  Represents the level of result (Optional). Must be one  of \code{ad}, 
#'  \code{adset}, \code{campaign}, \code{account}.
#' @param time.increment
#'  If it is an integer, it is the number of days from 1 to 90. 
#'  After you pick a reporting period by using \code{time.range} or 
#'  \code{date.preset}, you may choose to have the results for the whole
#'   period, or have results for smaller time slices. If "all_days" is used,
#'    it means one result set for the whole period. If "monthly" is used, 
#'    you will get one result set for each calendar month in the given period.
#'     Or you can have one result set for each N-day period specified by this
#'      param.
#' @param time.range
#'  time range must be \code{c(since = 'YYYY-MM-DD', until='YYYY-MM-DD')}
#' @param n
#'  Number of results to retrieve, defaults to \code{100}. 
#'  When you make an API request, you will usually not receive all of the 
#'  results of that request in a single response. 
#'  This is because some responses could contain thousands of objects so 
#'  most responses are paginated by default.
#' @param token
#'  A valid token as returned by \code{\link{fbAuthenticate}} or a 
#'  short-term token from 
#'  \href{https://developers.facebook.com/tools/explorer}{facebook Graph API Explorer}.
#' @param summary
#'  Default value: \code{FALSE}
#'  Determine whether to return a summary section with the same fields as 
#'  specified by fields will be included in the summary section. If 
#'  \code{TRUE} the data structure returned will be of type list 
#'  (see @return below)
#' @param verbose
#'  Defaults to \code{FALSE} if \code{TRUE} will print information on the 
#'  queries in the console.
#' @param limit
#'  Number of results requested at each API call, defaults to 100.
#'  Sometimes useful to bring it down if many results (\code{n}) are required as the 
#'  API might otherwise return \code{error_code: 1} or in other words an
#'   "Unknown error".
#' 
#' @details This function refers to the following API call \url{https://developers.facebook.com/docs/marketing-api/reference/ad-account/insights/},
#' it is strongly encouraged to have a look a the latter link.
#' only the following parameters are not available \code{default_summary}, 
#' \code{filtering}, \code{summary}, \code{sort} and \code{time_ranges}.
#' 
#' \code{getAny} may replace any other GET function. Indeed all functions share the same parameters and fields, the data that is returned solely depend on the object id the functions is fed. See examples.
#' 
#' @return Note that if summary is \code{TRUE} the function returns a 
#' \code{list}
#' 
#' @examples 
#' \dontrun{
#' # run authentication with your app details
#' fbOAuth <- fbAuthenticate(app.id = "1234567890123456", 
#'                           app.secret = "76xx79121xx0130x2x10a08x3e2x80xx", 
#'                           scope = "ads_management")
#'                           
#' # get ads on account
#' ads <- grabAds(id = "act_1014262775316748", token = TK)
#' 
#' # get adsets on account
#' adsets <- grabAdsets(id = "act_1014262775316748", token = TK)
#' 
#' # get ads on account
#' camp <- grabCampaigns(id = "act_1014262775316748", token = TK)
#' 
#' # get date.preset (this quarter)
#' date <- findParams("date.preset")[grep("quarter", findParams("date.preset"))]
#' 
#' # fetch Ad data
#' dat <- getAny(id = sample(ads$id, 1), token = TK, date.preset = date)
#' 
#' # fetch Adset data
#' dat <- getAny(id = sample(adsets$id, 1), token = TK, date.preset = date)
#' 
#' # fetch Campaign data
#' dat <- getAny(id = sample(campaigns$id, 1), token = TK, date.preset = date)
#' 
#' # fetch Account data
#' dat <- getAny(id = "act_1014262775316748", token = TK, date.preset = date)
#' }
#' 
#' @export
#' 
#' @seealso \code{\link{fbAuthenticate}}, \code{\link{findParams}},
#' \code{\link{grabAds}}, \code{\link{grabAdsets}}, 
#' \code{\link{grabCampaigns}} 
#' 
#' @author John Coene <john.coene@@cmcm.com>
getAny <- function(id, token, fields = "default", n = 100, 
                   action.attribution.windows, action.breakdowns, 
                   action.report.time, breakdowns, date.preset, level, 
                   time.increment, time.range, summary = FALSE,
                   verbose = FALSE, limit = 100) {
  
  # check arguments
  if(missing(action.attribution.windows)) action.attribution.windows <- NULL
  if(missing(action.breakdowns)) action.breakdowns <- NULL
  if(missing(action.report.time)) action.report.time <- NULL
  if(missing(breakdowns)) breakdowns <- NULL
  if(missing(date.preset)) date.preset <- NULL
  if(missing(level)) level <- NULL
  if(missing(time.increment)) time.increment <- NULL
  if(missing(time.range)) time.range <- NULL
  
  # check inputs
  if(missing(id)){
    stop("Missing id", call. = FALSE)
  }
  if (missing(token)){
    stop("Missing token", call. = FALSE)
  } 
  
  # check if region and action_carousel
  for (i in 1:length(fields)) {
    if(!is.null(breakdowns) && breakdowns == "region"){
      if(fields[i] == "action_carousel_card_id" ||
         fields[i] == "action_carousel_card_name"){
        stop(paste0("region cannot be used with action_carousel_card_id",
                    "or action_carousel_card_name"),
             call. = FALSE)
      }
    }
  }
  
  # create fields
  if(fields[1] == "default") {
    fields <- NULL
  } else {
    if(class(fields) != "character") {
      stop("Fields must be a character vector", 
           call. = FALSE)
    } else { 
      # test if fields correct
      testParam("fields", fields)
      
      # createFields
      fields <- createFields(fields)
    }
  }
  
  # action.attribution.windows
  if (length(action.attribution.windows)) {
    
    # test if correct
    testParam("action_attribution_windows", action.attribution.windows)
    
    # build action.attribution.windows
    action.attribution.windows <- paste0("&action_attribution_windows=",
                                         toHTTP(action.attribution.windows))
  }
  
  # action.breakdowns
  if (length(action.breakdowns)) {
    
    # test
    testParam("action_breakdowns", action.breakdowns)
    
    # build action.breakdowns
    action.breakdowns <- paste0("&action_breakdowns=", 
                                toHTTP(action.breakdowns))
  }
  
  # action.report.time
  if (length(action.report.time) == 1) {
    
    # test
    testParam("action_report_time", action.report.time)
    
    action.report.time <- paste0("&action_report_time=", action.report.time)
  } else if (length(action.report.time) > 1) {
    stop(paste0("action.report.time can only hold one value:", 
                "impression OR conversion"),
         call. = FALSE)
  }
  
  # breakdowns
  if(length(breakdowns)){
    breakdowns <- buildBreakdowns(breakdowns = breakdowns, f = testParam)
  }
  
  # date.preset
  
  # make date.preset NULL if time.range specified
  if (length(date.preset) == 1 && length(time.range) >= 1) {
    date.preset <- NULL
    warning("date.preset is ignored as time.range is specified",
            call. = FALSE)
  }
  
  if (length(date.preset) == 1) {
    
    # test
    testParam("date_preset", date.preset)
    
    date.preset <- paste0("&date_preset=", date.preset)
  } else if (length(date.preset) > 1) {
    stop(paste0("date.preset can only hold one of the following values: today,", 
                " yesterday, last_3_days, this_week, last_week, last_7_days,",
                " last_14_days, last_28_days, last_30_days, last_90_days, ",
                " this_month, last_month, this_quarter, last_3_months,",
                " lifetime"),
         call. = FALSE)
  }
  
  # level
  if (length(level) == 1) {
    
    # test
    testParam("level", level)
    
    level <- paste0("&level=", level)
  } else if (length(level) > 1) {
    stop(paste0("level can only hold one of the following values: ad, adset,",
                " campaign, account"), call. = FALSE)
  }  
  
  #time increment
  if(length(time.increment) == 1) {
    
    if(is.null(time.range) && is.null(date.preset)) {
      stop("time.increment must be used with either date.preset OR time.range",
           call. = FALSE)
    }
    
    # test if interger/numeric or character is passed
    if(class(time.increment) == "numeric" || class(time.increment) == "interger"){
      if(time.increment < 1 || time.increment > 90) {
        stop(paste0("If integer is passed to time.increment (number of days)",
                    " must be between 1 and 90. See @param."), call. = FALSE)
      } else {
        time.increment <- paste0("&time_increment=", time.increment)
      }
    } else {
      # test
      testParam("time_increment", time.increment)
      
      time.increment <- paste0("&time_increment=", time.increment)
    }
    
  } else if (length(time.increment) > 1) {
    stop(paste0("time.increment can only hold one of the following values:",
                " monthly OR all_days"), call. = FALSE)
  }
  
  # time range
  if(length(time.range) == 2) {
    
    date_check <- as.Date(time.range[2], format= "%Y-%m-%d")
    
    # further checks
    if(names(time.range)[1] != "since"){
      stop("time.range must be - c(since = 'YYYY-MM-DD', until = 'YYYY-MM-DD')",
           call. = FALSE)
    } else if (class(date_check) == "try-error" || is.na(date_check)){
      stop("Wrong date format. Must be YYYY-MM-DD", call. = FALSE)
    }
    
    time.range <- paste0("&time_range={'since':'", time.range[1], 
                         "','until':'", time.range[2], "'}")
    
  } else if (length(time.range) > 2) {
    stop("time.range must be - c(since = 'YYYY-MM-DD', until='YYYY-MM-DD')",
         call. = FALSE)
  }
  
  # summary
  summary <- paste0("&default_summary=", summary)
  
  # check token verison
  token <- checkToken(token)
  
  # build url
  uri <- paste0('https://graph.facebook.com/v2.7/',
                id, '/insights?fields=',fields,
                action.attribution.windows, action.breakdowns,
                action.report.time, breakdowns, date.preset, level,
                time.increment, time.range, summary,
                '&limit=', limit,'&access_token=', token)
  
  if(length(uri) > 1){
    stop("multiple ids supplied", call. = FALSE)
  }
  
  # call api
  response <- httr::GET(uri)
  
  # construct data
  fb_data <- constructFbAdsData(response)
  
  # parse data
  fb_data <- digest(fb_data)
  
  # paginate
  fb_data <- paginate(fb_data, n = n, verbose = verbose)
  
  # verbose
  if (verbose == TRUE) {
    message(paste(n, "results requested, API returned", nrow(fb_data$data), 
                  "rows"))
  }
  
  # converge
  fb_data <- converge(fb_data)
  
  if(class(fb_data) == "data.frame"){
    if (nrow(fb_data) == 0) warning(paste("No data."), call. = FALSE)
  }

  return(fb_data)
  
}